compiler/src/org.graalvm.compiler.core.aarch64/src/org/graalvm/compiler/core/aarch64/AArch64NodeMatchRules.java

```diff
@@ -233,38 +266,68 @@ public ComplexMatchResult extendedPointerAddShift(AArch64PointerAddNode addP) {
         return null;
     }
 
-    @MatchRule("(And (UnsignedRightShift=shift a Constant=b) Constant=c)")
-    @MatchRule("(LeftShift=shift (And a Constant=c) Constant=b)")
-    public ComplexMatchResult unsignedBitField(BinaryNode shift, ValueNode a, ConstantNode b, ConstantNode c) {
-        JavaKind srcKind = a.getStackKind();
-        assert srcKind.isNumericInteger();
-        AArch64BitFieldOp.BitFieldOpCode op = bitFieldOpMap.get(shift.getClass());
-        assert op != null;
-        int distance = b.asJavaConstant().asInt();
-        long mask = c.asJavaConstant().asLong();
+    @MatchRule("(And (UnsignedRightShift=shift value Constant=a) Constant=b)")
+    @MatchRule("(LeftShift=shift (And value Constant=b) Constant=a)")
+    public ComplexMatchResult unsignedBitField(BinaryNode shift, ValueNode value, ConstantNode a, ConstantNode b) {
+        JavaKind kind = shift.getStackKind();
+        return emitUnsignedBitField(kind, shift, value, a, b);
+    }
 
-        // The Java(R) Language Specification CHAPTER 15.19 Shift Operators says:
-        // "If the promoted type of the left-hand operand is int(long), then only the five(six)
-        // lowest-order bits of the right-hand operand are used as the shift distance."
-        distance = distance & (srcKind == JavaKind.Int ? 0x1f : 0x3f);
+    @MatchRule("(LeftShift=shift (ZeroExtend=extend (And value Constant=b)) Constant=a)")
+    @MatchRule("(ZeroExtend=extend (And (UnsignedRightShift=shift value Constant=a) Constant=b))")
+    @MatchRule("(ZeroExtend=extend (LeftShift=shift (And value Constant=b) Constant=a))")
+    public ComplexMatchResult unsignedExtBitField(ZeroExtendNode extend, BinaryNode shift, ValueNode value, ConstantNode a, ConstantNode b) {
+        JavaKind kind = extend.getStackKind();
+        return emitUnsignedBitField(kind, shift, value, a, b);
+    }
 
-        // Constraint 1: Mask plus one should be a power-of-2 integer.
-        if (!CodeUtil.isPowerOf2(mask + 1)) {
-            return null;
-        }
-        int width = CodeUtil.log2(mask + 1);
+    @MatchRule("(LeftShift=shift (SignExtend value) Constant)")
+    public ComplexMatchResult signedBitField(LeftShiftNode shift) {
+        JavaKind kind = shift.getStackKind();
+        assert kind.isNumericInteger();
+
+        SignExtendNode extend = (SignExtendNode) shift.getX();
+        int srcBits = extend.getInputBits();
+        int resultBits = extend.getResultBits();
+        assert kind.getBitCount() == resultBits;
+
+        int lsb = shift.getY().asJavaConstant().asInt() & (resultBits - 1);
```

Comment:
I think this is missing the mask of the lower bits in the shift amount. Here is a test:
```
diff --git a/compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java b/compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java
index 7212af4f80f..13ad08c91f8 100644
--- a/compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java
+++ b/compiler/src/org.graalvm.compiler.core.aarch64.test/src/org/graalvm/compiler/core/aarch64/test/AArch64BitFieldTest.java
@@ -183,6 +183,16 @@ public class AArch64BitFieldTest extends AArch64MatchRuleTest {
         testAndCheckLIR("signedB2LInsert", 0xab3213efL);
     }
 
+    public long signedB2LInsertShAmtOverflow(long input) {
+        byte b = (byte) input;
+        return ((long) b) << 64;
+    }
+
+    @Test
+    public void testSignedB2LInsertShAmtOverflow() {
+        testAndCheckLIR("signedB2LInsertShAmtOverflow", 0xab3213efL);
+    }
+
     // SBFIZ with S2L.
     public long signedS2LInsert(long input) {
         short s = (short) input;
```

Reply:
Left shift by `64` equals to `0`. So this case can use a signed extended instruction  
```
sxtb        x0, w2
```

Reply:
Ah. I see. I didn't pay too much attention to the output of the test. Yes this was false alarm. Everything looks good.